File: P1/ruby/Deepspace.rb
# encoding: UTF-8

module Deepspace
	module CombatResult
		ENEMYWINS = :enemywins
		NOCOMBAT = :nocombat
		STATIONSCAPES = :stationscapes
		STATIONSWINS = :stationswins
	end
	
	module GameCharacter
		ENEMYSTARSHIP = :enemystarship
		SPACESTATION = :spacestation
	end
	
	module ShotResult
		DONOTRESIST = :donotresist
		RESIST = :resists
	end
	
	module WeaponType
		class Type
			def initialize (p)
				@power = p
			end
			
			attr_reader :power # Observador de @power, equivalente a un método que lo devuelva
		end
		
		LASER = Type.new(2.0)
		MISSILE = Type.new(3.0)
		PLASMA = Type.new(4.0)
	end
	
	class Loot
		def initialize (nSu, nWe, nSh, nHa, nMe)
			@nSupplies = nSu
			@nWeapons = nWe
			@nShields = nSh
			@nHangars = nHa
			@nMedals = nMe
		end
		
		attr_reader :nSupplies
		attr_reader :nWeapons
		attr_reader :nShields
		attr_reader :nHangars
		attr_reader :nMedals
	end
	
	class SuppliesPackage
		def initialize (ap, fu, sp)
			@ammoPower = ap
			@fuelUnits = fu
			@shieldPower = sp
		end
		
		attr_reader :ammoPower
		attr_reader :fuelUnits
		attr_reader :shieldPower
		
		def self.newCopy (other) # Constructor de copia (es un método de clase)
			new(other.ammoPower, other.fuelUnits, other.shieldPower)
		end
	end
	
	class ShieldBooster
		def initialize (n, b, u)
			@name = n
			@boost = b
			@uses = u
		end
		
		attr_reader :name  # ¿Debería ser privado?
		attr_reader :boost
		attr_reader :uses
		
		def self.newCopy (other)
			new(other.name, other.boost, other.uses)
		end
		
		def useIt
			if @uses > 0
				@uses -= 1
				return @boost
			else
				return 1.0
			end
		end
	end
	
	class Weapon
		def initialize (n, t, u)
			@name = n
			@type = t
			@uses = u
		end
		
		attr_reader :name
		attr_reader :boost
		attr_reader :uses
		
		def self.newCopy (other)
			new(other.name, other.boost, other.uses)
		end
		
		def power
			@type.power
		end
		
		def useIt
			if @uses > 0
				@uses -= 1
				return power
			else
				return 1.0
			end
		end
	end
	
	class Dice
		def initialize
			@NHANGARSPROB = 0.25
			@NSHIELDSPROB = 0.25
			@NWEAPONSPROB = 0.33
			@FIRSTSHOTPROB = 0.5
		end
		
		# Para los siguientes métodos se usan las funciones rand
		# rand -> genera un número "real" entre 0.0 y 1.0
		# rand(x) -> genera un entero entre 0 y x (excluido x)
		
		def initWithNHangars
			rand < @NHANGARSPROB ? 0 : 1
		end
		
		def initWithNWeapons
			case rand
			when 0.0 .. @NWEAPONSPROB
				1	
			when @NWEAPONSPROB .. 2 * @NWEAPONSPROB
				2
			else
				3
			end
		end
		
		def initWithNShields
			rand < @NSHIELDSPROB ? 0 : 1
		end
		
		def whoStarts (nPlayers)
			rand(nPlayers)
		end
		
		def firstShot
			rand < @FIRSTSHOTPROB ? GameCharacter::SPACESTATION : GameCharacter::ENEMYSTARSHIP
		end
		
		def spaceStationMoves (speed)
			rand < speed
		end
	end

	class TestP1
		def self.main			
			una_instancia = WeaponType::Type.new(28)
			puts una_instancia.inspect
			puts una_instancia.power

			un_loot = Loot.new(1, 2, 3, 4, 5)
			puts un_loot.inspect
			#puts un_loot.nShields

			un_supplies_package = SuppliesPackage.new(1, 2, 3)
			puts un_supplies_package.inspect
			otro_supplies_package = SuppliesPackage.newCopy(un_supplies_package)
			puts otro_supplies_package.inspect

			un_shield_booster = ShieldBooster.new(1, 2, 0)
			puts un_shield_booster.inspect
			puts un_shield_booster.useIt
			otro_shield_booster = ShieldBooster.newCopy(un_shield_booster)
			puts otro_shield_booster.inspect

			my_weapon = Weapon.new("Nombrecillo", WeaponType::PLASMA, 0)
			puts my_weapon.inspect
			puts my_weapon.power
			puts my_weapon.useIt
			otra_weapon = Weapon.newCopy(my_weapon)
			puts otra_weapon.inspect

			my_dice = Dice.new
			puts my_dice.inspect
			puts my_dice.initWithNHangars
			puts my_dice.initWithNWeapons
			puts my_dice.initWithNShields
			puts my_dice.whoStarts(6)
			puts my_dice.firstShot
			puts my_dice.spaceStationMoves(0.75)
			
			mi_dice = Dice.new
			data = {:initWithNHangars => {0 => 0, 1 => 0}, :initWithNWeapons => {1 => 0, 2 => 0, 3 => 0}, :initWithNShields => {0 => 0, 1 => 0}, :whoStarts => {}, :firstShot => {GameCharacter::SPACESTATION => 0, GameCharacter::ENEMYSTARSHIP => 0}, :spaceStationMoves => {true => 0, false => 0}}
			# Completar data[:whoStarts] con las posibles claves
			
			ngamers = 6
			for i in 0..ngamers-1
				data[:whoStarts][i] = 0
			end
			
			# Calcular los datos de ejecución
			for i in 0..99
				data[:initWithNHangars][mi_dice.initWithNHangars] += 1
				data[:initWithNWeapons][mi_dice.initWithNWeapons] += 1
				data[:initWithNShields][mi_dice.initWithNShields] += 1
				data[:whoStarts][mi_dice.whoStarts(ngamers)] += 1
				data[:firstShot][mi_dice.firstShot] += 1
				data[:spaceStationMoves][mi_dice.spaceStationMoves(0.5)] += 1
			end
			
			puts "\nTras ejecutar cada método 100 veces, obtenemos lo siguiente: "
			for d in data
				print "Metodo "
				puts d
			end
		end
	end
end

Deepspace::TestP1.main

File: P1/java/deepspace/src/deepspace/Deepspace.java
package deepspace;

import java.util.function.Supplier;

enum CombatResult {ENEMYWINS, NOCOMBAT, STATIONESCAPES, STATIONWINS}
enum GameCharacter {ENEMYSTARSHIP, SPACESTATION}
enum ShotResult {DONOTRESIST, RESIST}

enum WeaponType {
    LASER (2.0f), MISSILE (3.0f), PLASMA (4.0f);
    private float power;
    
    WeaponType (float p) {
        power = p;
    }
    
    float getPower() {
        return power;
    }
}

class Loot {
    private int nSupplies, nWeapons, nShields, nHangars, nMedals;
    
    Loot (int nSu, int nWe, int nSh, int nHa, int nMe) {
        nSupplies = nSu;
        nWeapons = nWe;
        nShields = nSh;
        nHangars = nHa;
        nMedals = nMe;
    }
    
    int getNSupplies() {
        return nSupplies;
    }
    
    int getNWeapons() {
        return nWeapons;
    }
    
    int getNShields() {
        return nShields;
    }
    
    int getNHangars() {
        return nHangars;
    }
    
    int getNMedals() {
        return nMedals;
    }
}

class SuppliesPackage{
    private float ammoPower, fuelUnits, shieldPower;
    
    SuppliesPackage(float am, float fu, float sh) {
        ammoPower = am;
        fuelUnits = fu;
        shieldPower = sh;
    }
    
    SuppliesPackage(final SuppliesPackage other) {
        ammoPower = other.ammoPower;
        fuelUnits = other.fuelUnits;
        shieldPower = other.shieldPower;
    }
    
    float getAmmoPower() {
        return ammoPower;
    }
    
    float getFuelUnits() {
        return fuelUnits;
    }
    
    float getShieldPower() {
        return shieldPower;
    }
}

class ShieldBooster{
    private String name;
    private float boost;
    private int uses;
    
    ShieldBooster(String n, float b, int u) {
        name = n;
        boost = b;
        uses = u;
    }
    
    ShieldBooster(final ShieldBooster other) {
        name = other.name;
        boost = other.boost;
        uses = other.uses;
    }
    
    float getBoost(){
        return boost;
    }
    
    int getUses(){
        return uses;
    }
    
    float useIt() {
        if(uses > 0){
            uses -= 1;
            return boost;
        }else
            return 1;
    }
}

class Weapon{
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String n, final WeaponType w, int u) {
        name = n;
        type = w;
        uses = u;
    }
    
    Weapon(final Weapon other) {
        name = other.name;
        type = other.type;
        uses = other.uses;
    }
    
    WeaponType getType() {
        return type;
    }
    
    int getUses() {
        return uses;
    }
    
    float power() {
        return type.getPower();
    }
    
    float useIt() {
        if(uses > 0){
            uses -= 1;
            return power();
        }else
            return 1;
    }
}

class Dice{
    private final float NHANGARSPROB = (float)0.25;
    private final float NSHIELDSPROB = (float)0.25;
    private final float NWEAPONSPROB = (float)0.33;
    private final float FIRSTSHOTPROB = (float)0.5;
    private float generator = (float)Math.random();
    
    Dice() {}
    
    int initWithNHangars() {
        generator = (float)Math.random();
        if(generator < NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    
    int initWithNWeapons() {
        generator = (float)Math.random();
        if(generator < NWEAPONSPROB)
            return 1;
        else{
            if(generator < 2*NWEAPONSPROB)
                return 2;
            else
                return 3;
        }
    }
    
    int initWithNShields() {
        generator = (float)Math.random();
        if(generator < NSHIELDSPROB)
            return 0;
        else
            return 1;
    }
    
    int whoStarts(int nPlayers) {
        return (int)(Math.random()*nPlayers);
    }
    
    GameCharacter firstShot() {
        generator = (float)Math.random();
        if(generator < FIRSTSHOTPROB)
            return GameCharacter.SPACESTATION;
        else
            return GameCharacter.ENEMYSTARSHIP;
    }
    
    boolean spaceStationMoves(float speed) {
        generator = (float)Math.random();
        return generator < speed;
    }
}

public class Deepspace {
    public static void main(String[] args) {
        
        WeaponType wt = WeaponType.MISSILE;
        System.out.println("\nEl power del arma es: " + wt.getPower() + "\n");
        
        System.out.println("----------------------------------------------------\n");
        
        Loot l = new Loot(1,2,2,3,1);
        System.out.println("Los valores actuales del loot son: \n" + l.getNSupplies() + " " + l.getNWeapons() + " " + l.getNShields() + " " +  l.getNHangars() + " " + l.getNMedals() + "\n");
        
        System.out.println("----------------------------------------------------\n");
        
        SuppliesPackage s = new SuppliesPackage(2,5,10);
        System.out.println("Los valores actuales del paquete de suministros son: \n" + s.getAmmoPower()+ " " + s.getFuelUnits() + " " + s.getShieldPower());
        SuppliesPackage s_otro = new SuppliesPackage(s);
        System.out.println("Los valores actuales del otro paquete de suministros son: \n" + s_otro.getAmmoPower()+ " " + s_otro.getFuelUnits() + " " + s_otro.getShieldPower() + "\n");
        
        System.out.println("----------------------------------------------------\n");
        
        ShieldBooster sh = new ShieldBooster("Juan",5,2);
        System.out.println("Los valores actuales del protector de escudo son: \n" + sh.getBoost() + " " + sh.getUses());
        System.out.println("Al usar el protector: " + sh.useIt());
        ShieldBooster sh_otro = new ShieldBooster(sh);
        System.out.println("Los valores actuales del otro protector de escudo son: \n" + sh_otro.getBoost() + " " + sh_otro.getUses() + "\n");
        
        System.out.println("----------------------------------------------------\n");
        
        Weapon w = new Weapon("Alberto", wt, 5);
        System.out.println("El tipo de arma es: " + w.getType());
        System.out.println("Los usos del arma son: " + w.getUses());
        System.out.println("El poder del arma es: " + w.power());
        System.out.println("Al usar el arma: " + w.useIt() + "\n");
        Weapon w2 = new Weapon(w);
        System.out.println("El tipo de arma es: " + w2.getType());
        System.out.println("Los usos del arma son: " + w2.getUses());
        System.out.println("El poder del arma es: " + w2.power());
        System.out.println("Al usar el arma: " + w2.useIt() + "\n");
        
        System.out.println("----------------------------------------------------\n");
        
        Dice d = new Dice();
        System.out.println("Los hangares comienzan con el valor: " + d.initWithNHangars());
        System.out.println("Las armas comienzan con el valor: " + d.initWithNWeapons());
        System.out.println("Los escudos comienzan con el valor: " + d.initWithNShields());
        System.out.println("Comienza el jugador número " + (d.whoStarts(6)+1));
        System.out.println("El primer disparo lo ejecutará " + d.firstShot());
        if(d.spaceStationMoves((float)0.5))
            System.out.println("La estación ha esquivado el disparo\n");
        else
            System.out.println("La estación NO ha esquivado el disparo\n");
        
        System.out.println("----------------------------------------------------");
        System.out.println("-------------- LAS 100 EJECUCIONES -----------------");
        System.out.println("----------------------------------------------------\n");
        
        Dice di = new Dice();
        
        int []IWH = {0,0}; 
        int []IWW = {0,0,0};
        int []IWS = {0,0};
        int []players = {0,0,0,0,0,0};
        int []first = {0,0};
        int []SSM = {0,0};
       
        for(int i=0; i<100; i++){
            if(di.initWithNHangars() == 0)
                IWH[0]++;
            else
                IWH[1]++;
            
            if(di.initWithNWeapons() == 1)
                IWW[0]++;
            else{
                if(di.initWithNWeapons() == 2)
                    IWW[1]++;
                else
                    IWW[2]++;
            }
            
            if(di.initWithNShields() == 0)
                IWS[0]++;
            else
                IWS[1]++;
            
            switch(di.whoStarts(6)){
                case 0: players[0]++; break;
                case 1: players[1]++; break;
                case 2: players[2]++; break;
                case 3: players[3]++; break;
                case 4: players[4]++; break;
                case 5: players[5]++; break;
            }
            
            if(di.firstShot() == GameCharacter.SPACESTATION)
                first[0]++;
            else
                first[1]++;
            
            if(di.spaceStationMoves((float)0.5))
               SSM[0]++;
            else
               SSM[1]++;
        }
        
        System.out.println("Los valores de inicio de los hangares son:");
        for(int i=0; i<2; i++)
            System.out.println(IWH[i]);
        
        System.out.println("\nLos valores de inicio de las armas son:");
        for(int i=0; i<3; i++)
            System.out.println(IWW[i]);
        
        System.out.println("\nLos valores de inicio de los escudos son:");
        for(int i=0; i<2; i++)
            System.out.println(IWS[i]);
        
        System.out.println("");
        for(int i=0; i<6; i++)
            System.out.println("El jugador " + (i+1) + " comienza en " + players[i] + " ocasiones");
        
        System.out.println("\nEl equipo que comienza será: ");
        for(int i=0; i<2; i++)
            System.out.println(first[i]);
        
        System.out.println("\nLa estación se mueve en las siguientes situaciones: ");
        for(int i=0; i<2; i++)
            System.out.println(SSM[i]);
        
        System.out.println("");
    }
}